src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs

```diff
@@ -76,24 +93,30 @@ public partial class JsonTypeInfo<T>
                 Debug.Assert(CanUseSerializeHandler);
                 Debug.Assert(Converter is JsonMetadataServicesConverter<T>);
 
-                using var bufferWriter = new PooledByteBufferWriter(Options.DefaultBufferSize);
-                Utf8JsonWriter writer = Utf8JsonWriterCache.RentWriter(Options, bufferWriter);
+                Utf8JsonWriter writer = Utf8JsonWriterCache.RentWriter(Options, serializationContext.BufferWriter);
 
                 try
                 {
-                    SerializeHandler(writer, rootValue!);
-                    writer.Flush();
+                    try
+                    {
+                        SerializeHandler(writer, rootValue!);
+                        writer.Flush();
+                    }
+                    finally
+                    {
+                        // Record the serialization size in both successful and failed operations,
+                        // since we want to immediately opt out of the fast path if it exceeds the threshold.
+                        OnRootLevelAsyncSerializationCompleted(writer.BytesCommitted + writer.BytesPending);
```

Comment:
I'm thinking that it should be possible to collapse the two `finally` blocks into one. This can be done by moving the `OnRootLevelAsyncSerializationCompleted` call after the async flush operation (order shouldn't be significant in this case).

Reply:
The problem is FlushAsync could throw so we need the Dispose in a finally, but we also need to call ReturnWriter before any async code runs and the serialize code could throw so returning the writer needs to be in a separate finally so async doesn't mess with it.

Reply:
Good point, thanks.